src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/RoomService.kt

```diff
@@ -1,31 +1,57 @@
 package uk.gov.justice.digital.hmpps.approvedpremisesapi.service
 
+import org.springframework.data.repository.findByIdOrNull
 import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.ApprovedPremisesEntity
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.BedEntity
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.BedRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.CharacteristicRepository
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.PremisesEntity
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.RoomEntity
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.RoomRepository
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.TemporaryAccommodationPremisesEntity
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.validated
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.ValidatableActionResult
+import java.util.Optional
 import java.util.UUID
 
 @Service
 class RoomService(
   private val roomRepository: RoomRepository,
   private val bedRepository: BedRepository,
+  private val characteristicRepository: CharacteristicRepository,
 ) {
 
   fun createRoom(
     premises: PremisesEntity,
     roomName: String,
     notes: String?,
+    characteristics: List<UUID>,
   ): ValidatableActionResult<RoomEntity> = validated {
     if (roomName.isEmpty()) {
       "$.name" hasValidationError "empty"
     }
 
+    val characteristicEntities = characteristics.mapIndexed { index, uuid ->
+      val entity = characteristicRepository.findByIdOrNull(uuid)
+
+      if (entity == null) {
+        "$.characteristics[$index]" hasValidationError "doesNotExist"
+      } else {
+        if (entity.modelScope != "room") {
+          "$.characteristics[$index]" hasValidationError "incorrectCharacteristicModelScope"
+        }
+        if (premises is TemporaryAccommodationPremisesEntity && entity.serviceScope != "temporary-accommodation") {
+          "$.characteristics[$index]" hasValidationError "incorrectCharacteristicServiceScope"
+        }
+        if (premises is ApprovedPremisesEntity && entity.serviceScope != "approved-premises") {
+          "$.characteristics[$index]" hasValidationError "incorrectCharacteristicServiceScope"
+        }
```

Comment:
I was looking to see if these ifs can be consolidated into something a when but I don't think they can due to checking the state of both modelScope & serviceScope